File: solutions/go/grains/2/grains.go
package grains

import (
	"fmt"
	"math"
)

func Square(number int) (uint64, error) {
	if number <= 0 || number > 64 {
		return 0, fmt.Errorf("Square data is out of bounds")
	}
	return 1 << (uint64(number) - 1), nil
}

func Total() uint64 {
	return math.MaxUint64
}

File: solutions/swift/grains/2/Sources/Grains/Grains.swift
import Foundation

enum GrainsError: Error {
    case inputTooHigh(String)
    case inputTooLow(String)
}

private let minRange = 1
private let maxRange = 64
private let errorMessageGenerator: (Int) -> String = {
    return "Input[\($0)] invalid. Input should be between \(minRange) and \(maxRange) (inclusive)"
}

func square(_ number: Int) throws -> UInt64 {
    switch number {
    case ..<minRange:
        throw GrainsError.inputTooLow(errorMessageGenerator(number))
    case (maxRange + 1)...:
        throw GrainsError.inputTooHigh(errorMessageGenerator(number))
    default:
        return UInt64(pow(2.0, Double(number - 1)))
    }
}

var total: UInt64 {
    return (minRange...maxRange)
        .map { try? square($0) }
        .compactMap { $0 }
        .reduce(0, +)
}
